### ISSUE TITLE
rounding inconsistencies

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#8368]
This issue is to list inconsistencies in rounding numbers within the website and between the website and apps, and to discus which ones we want rounded up, or rounded down, or rounded off. It's not a complete list yet - if you notice any others, please comment here.

Items have been marked complete if they are correct or if an issue or PR has been created for them (even if the issue has not been fixed yet).

- Player's HP:
   - **desired rounding**: round **down** for health above 1 so that you don't think you have more health than you really have **but** we can't allow the situation where it looks like you have 0 health when you actually have slightly more than that because that produces bug reports ("I have zero health but I haven't died"), so round **up** for health less than 1
   - **desired decimal places**: zero decimal places for health greater than 1 but one decimal place for health less than 1.
  - examples of desired behaviour: 
      - 49.9 -> 49
      - 9.9 -> 9
      - 1.9 > 1
      - 0.99 -> 1
      - 0.9 -> 0.9
      - 0.19 -> 0.2
      - 0.09 -> 0.1
      - 0.0001 -> 0.1 
      - 0 -> 0
      - less than zero -> rounded down, showing negative sign and always zero decimal places (e.g., -0.1 -> -1)
  - [x] website: correct using this code: https://github.com/HabitRPG/habitica/blob/3b4b459e6861fd5e1c1abc5471fff95007559ff1/website/client/components/memberDetails.vue#L237-L243
  - [x] **Android app: rounds up** (49.9 -> 50, etc): https://github.com/HabitRPG/habitica-android/issues/914
  - [x] **Android widget: not tested, unlikely to be correct since app needs changes**: included in issue above
  - [ ] **iOS app: unsure, needs testing**: mentioned in Android issue above but a new issue should be created if needed

- Boss HP
  - **desired: rounds up** so that if you're aiming to kill the boss, you don't underestimate the damage you need to do.
      - However it would be nice if there was an on-hover effect on the website to show the actual value (I've seen two bug reports in the past couple of months related to this).
  - [ ] **website: truncates to two decimal places**  https://github.com/HabitRPG/habitica/pull/9749
  - [x] **Android: rounds down** https://github.com/HabitRPG/habitica-android/issues/915
  - [x] **iOS: rounds down** https://github.com/HabitRPG/habitica-ios/issues/442

- Pending damage to boss:
  - **desired**: shows one decimal place and **rounds down** so that if you're aiming to kill the boss, you don't overestimate the damage you have already done.
      - However it would be nice if there was an on-hover effect on the website to show the actual value (I've seen two bug reports in the past couple of months related to this).
     - The damage done to the boss perhaps could be more forgiving to avoid the situation where a player does 499.97 damage leaving a boss with 0.03 health - while accurate, that's not very motivating. As a specific example, when a user did 399.99945 damage to a boss with a starting HP of 400, the boss had 0.00055 health left but the website's party screen said the user had done 400.0 damage leaving the boss with 1 health, which was (rightly) reported as a bug.
   - [ ] **website: truncates to one decimal place** https://github.com/HabitRPG/habitica/pull/9749

- Player's Gold balance (e.g., 79.289): 
   - desired: round down to two decimal places (or truncate to two decimal places) so that you never try to buy something and are told that you don't have enough gold
   - [x] website: rounds down or truncates (79.28)
   - [x] Android app: rounds down/truncates (79.28)
   - [x] Android widget: rounds down/truncates (79.28)
   - [x] iOS app: floored as an integer (79)  https://github.com/HabitRPG/habitica-ios/issues/533

**Comment by Alys** [comment]
This issue is to list inconsistencies in rounding numbers within the website and between the website and apps, and to discus which ones we want rounded up, or rounded down, or rounded off. It's not a complete list yet - if you notice any others, please comment here.

- Player's HP (e.g., 33.02):
  - website: rounds up (34)
  - Android app: rounds up (34)
  - **iOS app: rounds down** (33) 
  - desired: round up so that when you have less than 1 health, it doesn't look like you have 0. Although when you have less than 10, it might be nice to start showing one decimal place since then the misleading effect of rounding up is more likely to lead to death. If we did that, we could start rounding off across the whole range.

- Boss HP
  - website: rounds up
  - Android: rounds up
  - iOS: _unknown_
  - desired: rounds up so that if you're aiming to kill the boss, you don't underestimate the damage you need to do, but it would be nice if there was an on-hover effect on the website to show the actual value (I've seen two bug reports in the past couple of months related to this)

**Comment by Alys** [comment]
I've updated the OP to say that on iOS, boss HP is rounded down.

**Comment by Alys** [comment]
It is possible (not confirmed) that the pending boss damage shown in the menu bar is rounded differently than the remaining boss health shown on the party page, leading to a player thinking that they'd done enough damage to kill the boss when they hadn't.

For example, this was reported in the Report a Bug guild by Bookwormy606 (1ea2c754-7f2b-4801-a815-6fb4f74a4f76): "It just happened that the damage pending on my menu bar was exactly the same as the boss health, but when I cronned it was .8 less than my pending."

**Comment by Alys** [comment]
From the Report a Bug guild by ReadyToBegin (31436f61-eac6-4886-8187-ebaa1d5080c6):
"Minor bug where widget for Android shows 1 less health than on site and on app. Ex. Widget has 47 when everything else has 48."

I've added that to the top post ("Android widget: rounds down")

**Comment by Alys** [comment]
From the Report a Bug guild by nickd (d5807fc1-af1d-49c1-94ec-5a613c18b18c): 
"A very minor inconsistency between the way gold is displayed in the Android App and the website. I don't think it matters, but thought I'd point it out just in case.I have 79.51924409624348 gold. The website says 79.51, presumably just displaying the first two characters after the decimal point. The android app says 79.52, presumably rounding up or down appropriately."

I've added that to the top post ("Player's Gold balance")

**Comment by majestic-owl448** [comment]
> Player's Gold balance (e.g., 79.5192): 
> - iOS app: unknown

Rigth now the value is floored as an integer (79). This since a couple of days after the removal of the silver coins (as first update there was a decimal value for gold like on the website)

**Comment by Alys** [comment]
@Ieahleen Thanks! I've updated the top post.

**Comment by Alys** [comment]
I'm going to do something to make this more actionable and/or actually action it.

**Comment by Alys** [comment]
I'm closing this because everything that still needs to be done is now documented elsewhere in GitHub (see the links in the top post).
